docs/ldml/tr35.md

```diff
@@ -2861,38 +2861,52 @@ Element content whose display may be affected in this way should include an expl
-| `range`        | <pre>= char ('-' char)?<br/>\| '{' (s char)+ s '}'</pre>       | a, a-c, \{abc}                          |
-| `prop`         | <pre>= '\\' [pP] '{' propName ([≠=] s value1+)? '}'<br/>\| '[:' '^'? propName ([≠=] s value2+)? ':]'</pre> | \\p\{x=y}, [:x=y:]<br/> |
-| `propName`     | <pre>= s [A-Za-z0-9] [A-Za-z0-9_\\x20]* s</pre>                | General_Category,<br/>General Category  |
-| `value1`       | <pre>= [^\\}]<br/>\| '\\' quoted</pre>                         | Lm,<br/>\\n,<br/>\\}                    |
-| `value2`       | <pre>= [^:]<br/>\| '\\' quoted</pre>                           | Lm,<br/>\\n,<br/>\\:                    |
-| `char`         | <pre>= [^\\& \\- \\[ \\[ \\] \\\\ \\} \\{ [:Pat_WS:]]<br/>\| '\\' quoted</pre> | a, b, c, \\n                            |
-| `quoted`       | <pre>= 'u' (hex{4} \| bracketedHex)<br/>\| 'x' (hex{2} \| bracketedHex)<br/>\| 'U00' ('0' hex{5} \| '10' hex{4})<br/>\| 'N{' propName '}'<br/>\| [[\u0000-\U00010FFFF]-[uxUN]]</pre> | _**error** if lengths not exact_ |
-| `charName`     | <pre>= s [A-Za-z0-9] [-A-Za-z0-9_\x20]* s</pre>                | TIBETAN LETTER -A                       |
-| `bracketedHex` | <pre>= '{' s hexCodePoint (s hexCodePoint)* s '}'</pre>        | \{61 2019 62}                           |
-| `hexCodePoint` | <pre>= hex{1,5} \| '10' hex{4}</pre>                           |                                         |
-| `hex`          | <pre>= [0-9A-Fa-f]</pre>                                       |                                         |
-| `s`            | <pre>= [:Pattern_White_Space:]*</pre>                          | optional whitespace                     |
-
-Some constraints on UnicodeSet syntax are not captured by this EBNF. Notably, property names and values are restricted to those supported by the implementation, and have additional constraints imposed by [[UAX44](https://www.unicode.org/reports/tr41/#UAX44)]. In addition, quoted values that resolve to more than one code point are disallowed in ranges of the form `char '-' char`.
+| `root`         | <pre>= prop<br/>\| '\[' '^'? s '-'? s seq\* \[\\$ \\-\]? s '\]' <br/>\| var</pre> | \\p\{x=y\},<br/>[abc],<br/>$myset                     |
+| `seq`          | <pre>= root \(s \[\\&\\\-\] s root\)\* s<br/>\| range s</pre>        | \[abc\]\-\[cde\], a                          |
+| `range`        | <pre>= element \('\-' element\)?       | a, a\-c, \{abc\}, \{ax\}\-\{bz\}                          | <!-- TODO: Support for string ranges of length != 1? -->
+| `prop`         | <pre>= '\\' \[pP\] '\{' propName \(\[≠=\] s value1\+\)? '\}'<br/>\| '\[:' '^'? propName \(\[≠=\] s value2\+\)? ':\]'</pre> | \\p\{x=y\}, \[:x=y:\]<br/> |
+| `propName`     | <pre>= s \[A\-Za\-z0\-9\] \[A\-Za\-z0\-9\_\\x20\]\* s</pre>                | General\_Category,<br/>General Category  |
+| `value1`       | <pre>= \[^\\\}\]<br/>\| '\\' quoted</pre>                         | Lm,<br/>\\n,<br/>\\\}                    |
```

Comment:
I opted for the slightly longer `pValuePerl` and `pValuePosix` to avoid confusion (does pvalueCurly _contain_ Curly or _exclude_ Curly?)

Reply:
I avoided touching the whole `prop` in general, because I don't quite understand that.

In a perfect world, there would be:
* An alternative nonterminal specifying the case where `propName` is left out (e.g., `[:L:]`
  * Reason being, what is the meaning of `[:≠:]`? Is it a syntax error, or `[:General_Category=≠:]`?
* An unambiguous set for `value1/2`:
  * whitespace is not excluded in `value`, but still valid syntax after `=/≠`
  * `\` is not excluded in `value`, so in theory both `[:name=\:]` and `[:name=\::]` are valid

I assume there's a reason for the value nonterminals to be so unrestrictive in the first place, so I feel underqualified to change them.

Reply:
> I assume there's a reason for the value nonterminals to be so unrestrictive in the first place

I am sure there is a reason, but I would not think there is a good one.

> what is the meaning of `[:≠:]`?

It is an error in ICU, as it should be, though not a syntax error. It might as well be a syntax error.

Note however that `[:meow:]` is also an error, but cannot be a syntax error (if we were to add a binary property `Meow`, a script `Meow`, or a General_Category alias `Meow`, it would become well-formed).

Reply:
Are there (supported) properties with "non-propName/charName" values? 
Is it valid for me to just search through https://www.unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt and  https://www.unicode.org/Public/UCD/latest/ucd/PropertyAliases.txt, or is there a more complete list?

I agree that we should not restrict UnicodeSet property _syntax_ to all currently-valid expressions of properties (`letter`, `letteR`, `lette r`, ...).

Reply:
- Perl and Posix are fine.
- If there is no propName it is a syntax error, so [::], [:=:], [:=L:], are all syntax errors (same for Perl syntax). As Robin says, we don't restrict the syntax to currently-valid properties. 

- Whitespace, _, case variants are allowed arbitrarily in propNames so [: d _ I _ :] = [:di:]
- As for the values, it depends. Most properties that behave like propName. [:gc=Lc:] = [: g c = L _ C :], [: n v = 1 _ 0 _ :] = [:nv=10:]
- String property values are different; whitespace (and _, and case differences) **are** meaningful. [:bmg=(:] ≠ [:bmg= ( :]
- We should reference the [the Regex spec](https://unicode.org/reports/tr18/) and in Section 5.9 Matching Rules of [[UAX44](https://www.unicode.org/reports/tr18/proposed.html#UAX44)] for propNames and pValues

I realized that there is another optional feature that we should describe here as well: implementations can add additional propNames and pValues.

Reply:
Are we defining how escaping should be handled here? For example, are we saying that `[:bmg=\::]` is referring to all code points with property `bmg` set to `:` (i.e., the `\:` is escaped)?

I haven't found a supported string property in ICU, so I don't know what the precedent is here.

My suggestion (assuming we want to handle escaping on the UnicodeSet side) would be something like:

```
prop = '[:' '^'? propName ([≠=] pValuePosix+)? ':]' ;
propName = [-A-Za-z0-9_[:Pattern_White_Space]]+ ; # Note: must contain at least one char in [A-Za-z0-9]
pValuePosix = [^ : \\ ] | '\' quoted ;
```

(and correspondingly for Perl property expressions)

This is compatible with [UAX44-LM3](https://www.unicode.org/reports/tr44/#UAX44-LM3) - implementations parse a propName that may contain whitespace, underscores, hyphens, and do the loose matching according to UAX44 themselves, if they want to. A similar thing for pValuePosix, except that escaping is supported, and whitespace may or may not be significant, hence the lack of a `Note` as in propName (it is significant in cases like `[:bmg= :]`, I assume).

(A side note on handling the escaping part in UnicodeSets: The playground does not escape, so `[:bmg=::]` is valid and results in `[\:]`, as is `[:bmg=\:]` which results in `[\\]`.)

I can try adding that (and your other comments about referencing UAX44 and custom propNames/pValues), if you agree

Reply:
https://util.unicode.org/UnicodeJsps/properties.jsp lists the properties with the value type (UCD, ICU,...)

You don't need to quote the : in your example, eg [[:Case_Folding=::]](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3ACase_Folding%3D%3A%3A%5D&g=&i=)

But you do need a quote in \p{Case_Folding=\}}

**Syntactically**, I don't think UnicodeSet in ICU currently cares what the character repertoire is for the propName or the pValues, other than the exclusion of =, ≠, } or :]. Although it is unlikely that the UCD would ever have non-ASCII, we might not want to limit it to ASCII. Please check the code for UnicodeSet to see what the constraints are.

Reply:
> https://util.unicode.org/UnicodeJsps/properties.jsp lists the properties with the value type (UCD, ICU,...)

`toNFM` is listed with source ICU, yet ICU throws an `Invalid name: toNFM` exception: https://github.com/unicode-org/icu/blob/70d308731ae01a4187fe49f7ae1c8e1ad1570be6/icu4j/main/classes/core/src/com/ibm/icu/lang/UCharacter.java#L5159

> You don't need to quote the : in your example, eg [[:Case_Folding=::]](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3ACase_Folding%3D%3A%3A%5D&g=&i=)

Exactly - this is invalid according to the `pValuePosix`/`value2` nonterminal. Do we want to support unquoted `:` characters in POSIX property sets, and add a worded-out note saying the sequence `:]` ends the expression eagerly?

> Please check the code for UnicodeSet to see what the constraints are.

`=` splits anything existing before `}` or `:]` into two. So yes, ICU currently does not care about the character repertoire. Interestingly, `≠` is not supported (neither is `!=`). Also, no escaping is happening. The code lives here: 
https://github.com/unicode-org/icu/blob/70d308731ae01a4187fe49f7ae1c8e1ad1570be6/icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java#L3797

Reply:
> yet ICU throws an Invalid name
The jsps add some custom properties. The ones that are supported in ICU (even if not in ICU's version of UnicodeSet) are still marked "ICU" in that list.

Re [[:Case_Folding=::]](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3ACase_Folding%3D%3A%3A%5D&g=&i=), I get what you mean not. Yes, that should be supported; probably best do with an explicit constraint note. 

Ok, ICU doesn't care about the repertoire. So there are three options, roughly:

1. XID based (but limited to ASCII), eg UCD style
2. XID based, eg adds other languages/scripts for future possible extension
3. Anything goes, eg current ICU

We could live with any of these (if not #3, we can always have a profile for that usage; if not #1, we can note that current props are always ASCII.) I guess I favor #2 as the cleanest, but it is not a strong preference. Others?

Reply:
No strong opinions from me here; I think I agree that 2 is cleaner, and it probably leads to better error messages (saying `[:]:]` or `[:cf!=fi:]` is a syntax error seems more helpful than complaining about `]` not being a known property or property value alias, or `cf!` not being a known property).

Reply:
> if not 3, we can always have a profile for that usage

Note that there is no need for a profile for 3, because the specification does not require distinguishing syntax errors from other errors: `[:≠:]` is just a bad UnicodeSet; ICU4C/J will tell you that is because `≠` is neither a binary property nor a GC or Script alias, and ICU4X might tell you that it is syntactically incorrect, but it is bad either way.